Math/NumberTheory/Primes/Factorisation/LinearAlgebra.hs

```diff
@@ -0,0 +1,257 @@
+module Math.NumberTheory.Primes.Factorisation.LinearAlgebra
+  ( SBVector(..)
+  , SBMatrix(..)
+  , mult
+  , size
+  , linearSolve
+  , testLinearSolver
+  ) where
+
+import qualified Data.List as L
+import qualified Data.IntSet as S
+import qualified Data.IntMap as I
+import qualified Data.Vector.Unboxed as U
+import Debug.Trace
+import Data.Semigroup()
+import System.Random
+import System.IO.Unsafe
+import GHC.Clock
+import Data.Bit
+import Data.Bits
+
+-- Sparse Binary Vector
+newtype SBVector = SBVector
+  { set :: S.IntSet
+  } deriving (Show, Eq)
+
+-- Vectors form a group (in particular a monoid) under addition.
+instance Semigroup SBVector where
+  SBVector v1 <> SBVector v2 = SBVector ((v1 S.\\ v2) <> (v2 S.\\ v1))
+
+instance Monoid SBVector where
+  mempty = SBVector mempty
+
+dot :: SBVector -> SBVector -> Bit
+dot (SBVector v1) (SBVector v2) = if even (S.size (v1 `S.intersection` v2)) then Bit True else Bit False
+-- Sparse Binary Matrix
+newtype SBMatrix = SBMatrix (I.IntMap SBVector) deriving (Show)
+
+mult :: SBMatrix -> SBVector -> SBVector
+mult (SBMatrix matrix) (SBVector vector) = foldMap (matrix I.!) (S.toList vector)
+
+size :: SBMatrix -> Int
+size (SBMatrix m) = I.size m
+
+linearSolve :: SBMatrix -> SBVector
+linearSolve matrix@(SBMatrix m) = linearSolveHelper 1 matrix randomVec
+  where
+    -- Make sure z is not empty.
+    randomVec = traceShowId $ SBVector (S.fromList $ randomSublist (I.keys m) (mkStdGen (fromIntegral (unsafePerformIO getMonotonicTimeNSec))))
+
+linearSolveHelper :: F2Poly -> SBMatrix -> SBVector -> SBVector
+linearSolveHelper previousPoly matrix z
+  -- In this case, z is in the image of matrix. Bad! Start again.
+  | null singularities          = trace ("Bad z.") $ linearSolve matrix
+  -- In this case potentialMinPoly is not a good approximation to minPoly
+  -- but vec is a good random vector.
+  | potentialSolution == mempty = trace ("Bad x.") $ linearSolveHelper potentialMinPoly matrix z
+  -- This is a good solution.
+  | otherwise                   = potentialSolution
+  where
+    potentialSolution = findSolution singularities matrix almostZeroVector
+    almostZeroVector = evaluate matrix z reducedMinPoly
+    (singularities, reducedMinPoly) = L.break (== Bit True) (U.toList $ unF2Poly potentialMinPoly)
+    -- lowest common multiple of previousPoly and candidateMinPoly CHECK
+    potentialMinPoly = lcm previousPoly candidateMinPoly
+    candidateMinPoly = findCandidatePoly matrix z
+
+findSolution :: [Bit] -> SBMatrix -> SBVector -> SBVector
+findSolution [] _ _ = mempty
+findSolution (_ : xs) matrix vector = if result == mempty then vector else findSolution xs matrix result
```

Comment:
It is better to use a guard here.
```haskell
findSolution (_ : xs) matrix vector 
  | result == mempty = vector
  | otherwise = findSolution xs matrix result
```